apiserver/facades/controller/charmdownloader/charmdownloader.go

```diff
@@ -107,7 +107,18 @@ func (a *CharmDownloaderAPI) downloadApplicationCharm(appTag names.ApplicationTa
 		return errors.Trace(err)
 	}
 
-	if !app.CharmPendingToBeDownloaded() {
+	resolvedOrigin := app.CharmOrigin()
+	if resolvedOrigin == nil {
+		return errors.NotFoundf("download charm for application %q; resolved origin", appTag.Name)
+	}
+
+	// In the case of deploying multiple applications utilizing the
+	// same charm, keep going to allow DownloadAndStore to return
+	// the correct origin to be saved below. The charm will not
+	// actually be downloaded more than once. The method will just
+	// provide the correct origin. Necessary for deploying resources
+	// and refreshing charms.
+	if !app.CharmPendingToBeDownloaded() && resolvedOrigin.ID != "" {
```

Comment:
I think we also need to check the resolveOrigin source, in incase it's a local charm?

Reply:
Local charms cannot be put in CharmPendingToBeDownloaded state, there is a check in state.